accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+
+#### Verifying
+The package integrity captured by the provenance attestation will be verified by the npm CLI. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.
+
+This will be staged rollout over several npm CLI releases as changing the default behavior of `npm install` is a breaking change:
+
+1. The existing `npm audit signatures` command will start verifying build signatures, possibly adding an opt-in flag to audit these signatures during `npm install`, e.g. `--audit-build-provenance`
+2. Start auditing build signatures by default when running `npm install`. This will only check packages that have build attestations and ignore the rest. Allow opting out of this behavior with `--no-audit-build-provenance` or `--no-audit`.
+3. Allow users to opt-in to a strict mode to require packages to be published with build provenance information, e.g. `npm install --require-build-provenance`. Initially, this might only be usable by maintainers that only depend on their dependencies. It will take a long time for this to be feasible for the average install as all transitive dependencies would need to include build provenance on the installed version. We could allow policies that make gradual enforcement possible, e.g. only require build provenance on packages published after some minimum date.
+
+Signature verification needs to be performant and have a negligible impact on `npm install`. We believe the main bottleneck here will be the network, downloading the signatures to verify. The crypto operations involved in verification are negligible on modern hardware. We propose the CLI makes a bulk API request to the npm registry, similar to how advisories are fetched, for all provenance attestations for a given install.
+
+##### Detailed steps to verify
+1. npm CLI downloads the provenance and release attestations for each installed package (the download will be done in bulk for all installed packages)
+    1. Verify the build provenance attestation is signed by a certificate created for the builder/actions run.
+    2. One Rekor entry proves that the build provenance attestation was placed on the public ledger.
+    3. One Rekor entry proves that the npm package was accepted and released by the npm registry (the Rekor entry contains the package’s url (PURL), sha512 digest and is signed with a certificate for the npm registry).
+    4. Verify both of the certificates are issued by Fulcio and identifying different workloads (builder and the npm registry).
+2. npm CLI verifies that the Rekor entries, certificate and certificate chain in the build provenance are trusted. The [sigstore-js](https://github.com/sigstore/sigstore-js) client includes a TUF client that's used to verify, and optionally update, the public keys for both Rekor and Fulcio.
```

Comment:
Can you also verify that the Fulcio certificate was issued to a certificate transparency log? Like WebPKI, Fulcio embeds a Signed Certificate Timestamp (SCT) in the certificate as a promise of inclusion in a log.

There is a separate key in the TUF targets for this.

Reply:
Yes I believe we're working on verifying the SCT in https://github.com/sigstore/sigstore-js (there's a bunch of verification steps missing in there atm)